src/common/browser-adapters/browser-adapter-event-manager.ts

```diff
@@ -14,115 +13,150 @@ export interface EventDetails {
-                ? this.timeoutFactory.createTimeout(
-                      () => Promise.resolve(result),
-                      TWO_MINUTES,
-                      `wrapper-timeout-${Date.now()}`,
-                  )
-                : this.timeoutFactory.createTimeout(() => Promise.resolve(result), TWO_MINUTES);
-        }
-        return result;
-    }
-
-    public processEvent(eventType: string, eventArgs: any[], isDeferred?: boolean): boolean {
-        if (this.eventsToApplicationListenersMapping[eventType]) {
-            this.eventsToApplicationListenersMapping[eventType].forEach(applicationListener => {
-                this.forwardEventToApplicationListener(applicationListener, eventArgs);
+            return this.promiseFactory.timeout(result, FOUR_MINUTES).catch(() => {
+                // We want to avoid rejecting the Promise we give back to the browser event handler
+                // because doing so is liable to cause the browser to tear down our Service Worker
+                // immediately, even if other events are in progress.
+                return undefined;
             });
-            return true;
         } else {
-            if (!isDeferred) {
-                const eventDetails: EventDetails = {
-                    eventType,
+            if (result === undefined) {
+                // it is possible that this is the result of a fire and forget listener
+                // wrap promise resolution in 2-minute timeout to ensure it completes during service worker lifetime
+                return this.promiseFactory.delay(result, TWO_MINUTES);
+            } else {
+                //this is an odd case we should be aware of.
+                //if triggered, update listener to return a Promise with the result
+                this.logger.error(
+                    `Application listener returned a non-promise result`,
+                    listener,
                     eventArgs,
-                };
-                this.deferredEvents.push(eventDetails);
+                );
+                return Promise.resolve(result);
             }
-            return false;
         }
     }
 
-    private unregisterAdapterListener(event: Events.Event<any>, eventType: string) {
-        event.removeListener(this.adapterListener(eventType));
-    }
+    private deferEvent(eventDetails: EventDetails): Promise<any> {
+        const deferredResolution = this.promiseFactory.externalResolutionPromise();
 
-    private unregisterApplicationListenerForEvent(
-        eventType: string,
-        listener: ApplicationListener,
-    ) {
-        const allListeners = this.eventsToApplicationListenersMapping[eventType];
+        this.deferredEvents.push({
+            deferredResolution: deferredResolution,
+            ...eventDetails,
+        });
+        //resolve the promise in four minutes in case a listener is never registered
+        setTimeout(() => {
+            deferredResolution.resolveHook('no listener registered');
```

Comment:
I think it would be good for us to log an error in this case as well; I don't think this is expected in any scenario